versions/draft-3/SPEC.md

```diff
@@ -3053,6 +3056,44 @@ Array[String] zs = [ "d", "e" ]
 Array[Pair[Int, String]] crossed = cross(xs, zs) # i.e. crossed = [ (1, "d"), (1, "e"), (2, "d"), (2, "e"), (3, "d"), (3, "e") ]
 ```
 
+## Array[Pair[X,Y]] as_pairs(Map[X,Y])
+
+Given a Map, the `as_pairs` function returns an Array containing each element in the form of a Pair. The key will be the left element of the Pair and the value the right element. The order of the the Pairs in the resulting Array is the same as the order of the key/value pairs in the Map.
+
+```
+Map[String, Int] x = {"a": 1, "b": 2, "c": 3}
+Map[String,Pair[File,File] y = {"a": ("a.bam, "a.bai"), "b": ("b.bam", "b.bai")}
+
+Array[Pair[String,Int]] xpairs = as_pairs(x) # [("a", 1), ("b", 2), ("c", 3)]
+Array[Pair[String,Pair[File,File]] ypairs = as_pairs(y) # [("a", ("a.bam", "a.bai")), ("b", ("b.bam", "b.bai"))]
+```
+
+## Map[X,Y] as_map(Array[Pair[X,Y]])
+
+Given an Array consisting of Pairs, the `as_map` function returns a Map in which the left elements of the Pairs are the keys and the right elements the values. The order of the key/value pairs in the resulting Map is the same as the order of the Pairs in the Array.
+
+In cases where multiple Pairs would produce the same key, the workflow will fail.
```

Comment:
I believe the keys to maps must be primitives whereas the left side of pairs can be arbitrarily complex. Maybe we should also add that as a note alongside this comment? 
```md
In cases where the left side of the pair is not a primitive type, the workflow will fail
```

Reply:
In fact... since hopefully we'll be able to detect this statically 99% of the time it should probably just be:
```
"The left hand side of the pair passed to `as_map` must be a primitive type"
```